### PR TITLE
Add Google BigQuery sink connector configuration

### PR DESCRIPTION
**Pull request by upendra-vedullapalli**
- created a terraform module that can be easily imported into your terraform infrastructure
- created an example terraform file depicting the usage of the module showing the minimal code required to configure while deploy and manage a connector
- Modular approach would ensure
  - correct Aiven provider version being used
  - required parameters for a connector being overridden

Notes:- 
1. Plan is to extend the list of modules with examples for all the supported connectors towards Aiven KafkaConnect cluster
2. Need to setup GitHub actions to validate and versioning of the module

### CODE REVIEW

**Review by maltel — APPROVED** [review]
Har ikke noe å utsette her 👍

**Comment by upendra-vedullapalli** [comment]
Noen fra Platform teamet er interessert å komentere på dette PR?

File: README.md [hunk 1]
```diff
@@ -1,2 +1,12 @@
-# terraform-aiven-kafka-connect
-Terraform module(s) for Aiven Kafka Connect
+# Terraform module(s) for Aiven Kafka Connect
+Modules for provisioning various Kafka connectors onto an Aiven managed KafkaConnect cluster
+
+## Getting started
+Access to Aiven terraform provider requires an API authentication token which can be generated from [Aiven console](https://console.gcp.aiven.io/profile/auth)
+The owner of the API Authentication token should have admin permissions.
+Export the token in an environment variable `TF_VAR_api_token`
```
Comment (williampeer) [inline README.md]: Should we perhaps reformulate this to something like (the variable) "needs to be supplied during provisioning, and may also be supplied in a .tfvars-file, or as a `TF_VAR_`-prefixed environment variable (...)"?

File: examples/bigquery-sink/main.tf [hunk 2]
```diff
@@ -0,0 +1,15 @@
+variable "api_token" {}
+module "bigquery-sink" {
+  # This is for local reference only; if you're using this module as a published
+  # module from GitHub, the 'source' parameter must refer to it's public location.
+  #  source                      = "github.com/entur/terraform-aiven-kafka-connect/modules/bigquery-sink?ref=vVERSION"
+  source                      = "../../modules/bigquery-sink"
+  aiven_api_token             = var.api_token
+  aiven_project_name          = "ent-sbx"
```
Comment (williampeer) [inline examples/bigquery-sink/main.tf]: hard-coded project - refactor into a variable with "ent-sbx" as the default value?

File: examples/bigquery-sink/main.tf [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  kafka_service_name          = "entur-kafka-test-int"
```
Comment (williampeer) [inline examples/bigquery-sink/main.tf:9]: same as above

File: examples/bigquery-sink/main.tf [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  bigquery_project_name       = "entur-data-raw"
+  bigquery_dataset_name       = "aiven_bq_sink_connector_module_test"
+  bigquery_service_account_id = "kcbqs-api"
+  kafka_topics                = ["test-topic-staging"]
+  connector_name              = "aiven-bq-sink-connector-module-test"
```
Comment (williampeer) [inline examples/bigquery-sink/main.tf:14]: same in general for the module

File: modules/bigquery-sink/README.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+| key_source_type                          | Key type for Google credentials with access to `bigquery_dataset_name`      |    string    |                       "JSON"                       |    no    |
+| key_converter                            | Converter class for key data on a connect record                            |    string    | "org.apache.kafka.connect.storage.StringConverter" |    no    |
+| value_converter                          | Converter class for value data on a connect record                          |    string    |     "io.confluent.connect.avro.AvroConverter"      |    no    |
+
+[Example](./../../examples/bigquery-sink/main.tf)
+Sample configuration for reference
+
+## Getting started
+
+- Set an environment variable `TF_VAR_api_token` with Aiven authentication token
```
Comment (williampeer) [inline modules/bigquery-sink/README.md]: See README-comment above

File: modules/bigquery-sink/connector.tf [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  account_id = var.bigquery_service_account_id
+  project    = var.bigquery_project_name
+}
+
+resource "google_service_account_key" "bq-sa-key" {
+  service_account_id = data.google_service_account.bq-sa.name
+}
+
+locals {
+  schema_registry_url      = "https://${data.aiven_service_component.schema-registry.host}:23714"
```
Comment (williampeer) [inline modules/bigquery-sink/connector.tf]: consider setting the port to a variable with a default value `23714`

File: examples/bigquery-sink/main.tf (hunk 2 above)
Comment (upendra-vedullapalli) [reply examples/bigquery-sink/main.tf]: This is an example configuration file depicting usage of the module. We could use more generic strings here.

Comment (williampeer) [reply examples/bigquery-sink/main.tf]: ah, I missed the `examples/`! 
then it makes sense with the current example

**Comment by AlexanderBrevig** [comment]
Seems to me there are _many_ inputs to this module. Are they all likely to change between the applications owned by a team in the same environment? Maybe we could use the [GCP 2.0 init module](https://github.com/entur/terraform-google-init) and infer (some/all of) these values?



For instance, the simplest call-site for creating a bucket looks like this:

https://github.com/entur/terraform-google-cloud-storage/blob/69442b69b17471e19f0758cdf326522457b74c2b/examples/minimal/main.tf#L11-L14

The goal should be to only require input that is actually under developer control. In my mind, a developer should really struggle to input something that will not work.

PS: I don't know this domain, and it may very well be that my suggestion is a stupid one. These are just my intuitive 0b10.

**Comment by upendra-vedullapalli** [comment]
> 

Thanks for the feedback @AlexanderBrevig 
Since these connectors are going to be deployed in Aiven's managed GCP cluster(s) we may not be able to use the same GCP2.0 init module. 
But I definitely see some properties going to be common to lot of connectors we plan to deploy.
So I can create a separate init module for Aiven KafkaConnect with common configuration and it can then be used in all other specific connector modules.

I also wanted to confirm from you guys if we have to create separate git repos for all terraform modules to be able to version them or we can achieve versioning if we host all modules in same repo?

**Comment by sindrel** [comment]
What are your plans for Aiven API authentication and IAM, i.e. how do you plan to issue API tokens? Have you found a way to segment access on a more specific scope than project level?

**Comment by AlexanderBrevig** [comment]
Because of gcp2 naming conventions, you should be able to infer a lot based on the existing init container. If you also have a convention for your clusters it should be easy to map init values to the correct values? Again, I just want the upstream terraform to be as clean and easily maintainable as possible.

You should definitely have a repo per module, take a look how things are set up with the other gcp2 modules. Should be reusable for you :)

**Comment by upendra-vedullapalli** [comment]
> What are your plans for Aiven API authentication and IAM, i.e. how do you plan to issue API tokens? Have you found a way to segment access on a more specific scope than project level?

Access to Aiven API is through access tokens specific to a [project member.](https://docs.aiven.io/docs/platform/concepts/projects_accounts_access.html) Aiven's terraform provider supports managing project members with e-mail addresses but does not support managing access tokens. Issue is reported to Aiven. So the user has to manage his/her token from [Aiven's web console](https://docs.aiven.io/docs/platform/howto/create_authentication_token.html). Project user access can be scoped at creation time with assigned role.
When it comes to IAM access to BigQuery projects/datasets from a connector is through service accounts. User deploying such connectors should have BigQuery admin privileges on that project to be able to manage service account keys.

**Pull request merged**